### ISSUE TITLE
Inline Styles?

### ISSUE THREAD

**Issue by nelsonic** [dwyl/learn-react#14]
Facebook *officially* recommends that people define their
style as a JavaScript Object: http://facebook.github.io/react/tips/inline-styles.html 
e.g:
```js
var divStyle = {
  color: 'white',
  backgroundImage: 'url(' + imgUrl + ')',
  WebkitTransition: 'all', // note the capital 'W' here
  msTransition: 'all' // 'ms' is the only lowercase vendor prefix
};

ReactDOM.render(<div style={divStyle}>Hello World!</div>, mountNode);
```
Does anybody else think that in-lining styles as JS inside Components is a ***debugging nightmare*** waiting to happen (*as an app gets larger and new people add hacks & overrides to mask previous CSS instead of editing the CSS in the style sheet...?

**Comment by jkbits1** [comment]
Interesting, it does look like a problem just waiting to happen. I'd definitely like to know if they have an argument why that's not the case.

**Comment by iteles** [comment]
This was exactly what put me off React in the first place. I would love to get some more views on this. @nelsonic Worth sharing the question on twitter?

**Comment by nelsonic** [comment]
@iteles nah, twitter is mostly people who are relatively *inexperienced* with large-scale react apps (*e.g: have only built a handful of few components - still maintainable*) and thus still think mixing style inside JS is a *great idea* ... and *nobody* at *Facebook* is going to badmouth *anything* about React because its their main [front-end-developer recruitment tool](https://www.pandastrike.com/posts/20150311-react-bad-idea) ... 

we need to survey people who are *attempting* to ***maintain*** a React web app that was built & deployed by *someone else* ...

I have *personal experience* of attempting to *Fix* other people's poorly implemented React components (*no names*) ... but I don't know if my *experience* is "*anecdotal*" or if others have worked on apps that are *large* enough (*and built by enough different people with insufficient testing*) to be *un-maintainable* ...

**Comment by iteles** [comment]
This is actually an interesting article that _doesn't_ advocate inline styling but _does_ go some of the way to explaining the reasoning for this to be part of react in the first place: https://medium.com/@ajsharp/please-please-don-t-use-css-in-js-ffeae26f20f#.pypeemetl

**Comment by iteles** [comment]
http://red-badger.com/blog/2015/11/19/london-react-meetup-november-2015/
Has a talk on inline styles (that I haven't watched yet, but want to keep it somewhere so I don't forget!)

**Comment by newswim** [comment]
Dunno if anyone has seen this yet, but looks cool ---> https://github.com/rofrischmann/fela

Based on 'atomic' css, a la http://acss.io (i think)

**Comment by nelsonic** [comment]
Hi @newswim `fela` looks _really_ promising as a CSS-in-JS tool! 
Great docs, tiny size and feature-rich. 👍 

We are have decided to use `Tachyons` on the basis that it takes the concept of "_in-line-styles_" _all-the-way_ to the "***Next Level***". i.e: has ***zero cascading*** and all style application is ***100 explicit*** and _targeted_ to the component being styled. It's ***impossible to break*** on any _other_ component by changing a style in a _different_ component. 

We _know_ it's not _everyone's_ "_cup-of-tea_" because _ironically_ it requires some CSS knowledge (_even though you are not using CSS **directly** when you use Tachyons you still need to know about Margin, Padding, Positioning, etc. in order to **use** the library..._)
See: https://github.com/dwyl/learn-tachyons 

Where "Atomic" CSS is designed for "Maximum Re-use", see: https://github.com/acss-io/atomizer
Tachyons is designed to give you the "_full power_" of CSS without ever writing a _line_ of CSS.
We can create an entire application simply by _composing_ Tachyons properties without ever having to create a `style.css` file or doing an "_pre-processing_" or thinking about setting up "_break points_" for "_responsiveness_". That is _literally_ ***all*** `done` for us which **saves an _immense_ amount of time** both in "_setup_" for a `new` project and when our team members are switching _between_ projects. ⏳ 

There are still _plenty_ of use cases for the _various_ CSS Frameworks/Libraries/Systems,
we are just really _loving_ Tachyons right now, because we've felt the _pain_ of using a more "_advanced_" framework and that requires a "_build tool_" to "_compile_" the CSS in our components and the perils of attempting to re-use styles which then get "_broken_" of "_over-ridden_" by well-meaning new team members who didn't know the style was used in more than one place ... 😢 

Anyway, thanks for sharing the link to `fela`! 
keep up the sharing it's always good to know about promising new tools! 👍 
❤️  ✅

**Comment by pankajpatel** [comment]
Except for JS based animations, I would never write CSS in JS.

We have LESS/SCSS with PostCSS for that.

I think it in terms of operation that Browser needs to perform to apply any change. And when a Style is added by JS, there is DOM manipulation which puts all the core parts of browser to work for each step.

With CSS, these operations happen in less number (I assume).

**Comment by nelsonic** [comment]
@pankajpatel _exactly_. 
Using Tachyons is _benchmarked_ as being (_much_) _faster_ on Mobile Clients than using JS to apply styles via "Virtual DOM". ***Performance*** is a "***core principal***" of Tachyons: http://tachyons.io/#principles

That is not to say that we don't _apply_ styles in our `elm` components when required.
Simply that:
a) we don't "_mutate_" the DOM/Styles in multiple places
b) the Tachyons classes are all _declarative_ no _magic methods_ which can be over-ridden somewhere.
c) tracking changes to the styles for an individual HTML Element or Component is _trivial_ which promotes accountability for the _exact_ "_look and feel_" for that Element/Component because it's not possible to have "***side effects***" when changing one element.

We will be producing an intro video to Tachyons 🔜  https://github.com/dwyl/the-book/issues/69 (_watch this space_)

Thanks again for _unearthing_ this issue/discussion thread! 👍

**Comment by pankajpatel** [comment]
Thanks @nelsonic for info about Tachyons. 

This seems to be a good fit for kick starting new projects. I found many kickstart projects for React and many were good fit, but I have to do the CSS setup most of the times. So I created one just for myself https://github.com/pankajpatel/kickstart-react.

It allowed me to choose between LESS and SCSS with just config file , but I thing I will add Tachyons as primary one in there.

**Comment by pankajpatel** [comment]
I came across a thread about the Browser Performance in Mobile devices, which primarily  stated that JS animations on mobile screens will lag because as part of optimising Battery Utilisation, browser pauses the JS execution while scrolling whereas CSS isn't. It sounded fair too for mobile browsers.

http://stackoverflow.com/questions/15871942/how-do-browsers-pause-change-javascript-when-tab-or-window-is-not-active

And now I'm thinking what would happen to CSS in JS.

**Comment by newswim** [comment]
@pankajpatel That's interesting! I hadn't at all thought of the introduction of lag due to JS runtime behavior on mobile. Thank you's to yourself and @nelsonic for the illuminating discussion.

I stumbled across this [how-to-style-react](http://andrewhfarmer.com/how-to-style-react/) article yesterday and felt utterly lost. It seems like `mxstbr`, one of the core React contributors has rallied some _[support](https://medium.freecodecamp.com/a-5-minute-intro-to-styled-components-41f40eb7cd55#.1dsvgipa9)_ behind **[styled-components](https://medium.freecodecamp.com/a-5-minute-intro-to-styled-components-41f40eb7cd55#.1dsvgipa9)**, but after reading all of the cross-talk on various threads I'm still at a bit of a loss.

Picking a library seems conditional on the use case, but all the divergent opinions (not from you folks) seems a bit at odds with the whole point of a front-end framework. While I'd like to use SCSS for all of the familar libs and functions, I can definitely see the gains from an extremely declarative library like Tachyons. Here it goes... 🏊

**Comment by newswim** [comment]
Styled-Components are about to published a _really cool_ utility library called [Polished](https://github.com/styled-components/polished).

If anyone used Stylus and remembers Rucksack or Nib, it reminds me of similar mixins. It seems to play very nicely with the larger Styled Components workflow. I know DWYL is partial to Tachyons, but this is one of the few React styling libraries that looks like it will be successful in the long term.

**Comment by nelsonic** [comment]
@newswim it's _always_ good to be _aware_ of what is _available_. thanks for sharing it! 👍 
https://github.com/styled-components/polished looks _promising_ 👀  🚀 
I remain _curious_ about ***all*** new technologies and approaches to solving common problems.

_However_ ... I think `polished` is _solving a different_ "_problem_" from the "_Functional CSS_" libraries.

I'm not an "expert" but ... each time I see a style library that requires ES6 I cannot help but feel it is (_almost by definition_) _over-complicating_ UI rendering in the pursuit of "_moar features_". 
If those ES6-enriched styles are pre-computed ("_transpiled_") by babel they _might_ be _fast_ to render, but I'm a _huge_ fan of the "_Less is More_" approach taken by Tachyons. 😍 

_More importantly_ what are _you_/your team using for your UI and _why_? 🤔

**Comment by pankajpatel** [comment]
The [article by Addy Osmani](https://medium.com/dev-channel/javascript-start-up-performance-69200f43b201#.bieanzsg9) published few days back explains clearly the peroformance in JS frameworks.

What would happen if CSS also becomes part of that initial loading? More time needed to respond after initial load.

BEM and similar methodologies already are there to easily manage the CSS' meaningfulness regarding the components.

**Comment by newswim** [comment]
@nelsonic my team just started using a combination of `styled-components` _and_ Tachyons, and will probably add Polished once it hits 1.0

The syntax is a little daunting at first (literally put all these together today), but it should help greatly with _creating a theme_ (the **why** question). 😸

**Comment by nelsonic** [comment]
https://zendev.com/2017/09/11/css-in-js.html
